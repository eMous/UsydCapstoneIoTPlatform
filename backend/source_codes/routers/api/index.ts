export {};
const express = require("express");
const rootRouter = express.Router();
const projectOwner = require("./ProjectOwner");
const participant = require("./Participant");

rootRouter.use("/projectowner", projectOwner);
rootRouter.use("/participant", participant);

module.exports = rootRouter;
