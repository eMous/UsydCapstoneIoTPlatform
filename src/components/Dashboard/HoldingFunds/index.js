import useTheme from "@material-ui/styles/useTheme";
import DataCard from "../components/DataCard";
import DataCardIcon from "../components/DataCardIcon";

const HoldingFunds = ({ fundsAmt }) => {
  const theme = useTheme();
  const holdingFundsIcon = (
    <DataCardIcon
      iconName="account_balance"
      iconBackgroundColour={theme.palette.secondary.main}
    />
  );

  const formatter = new Intl.NumberFormat("en-AU", {
    style: "currency",
    currency: "AUD",
  });

  return (
    <DataCard
      title="Funds in Holding (To be paid)"
      dataValue={formatter.format(fundsAmt)}
      displayIcon={holdingFundsIcon}
    />
  );
};

export default HoldingFunds;
