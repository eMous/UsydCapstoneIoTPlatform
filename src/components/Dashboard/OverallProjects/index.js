import { Link } from "react-router-dom";
import useTheme from "@material-ui/styles/useTheme";
import DataCard from "../components/DataCard";
import DataCardIcon from "../components/DataCardIcon";
import routes from "routes";

const OverallProjects = ({ numProjs }) => {
  const theme = useTheme();
  const overallProjsIcon = (
    <DataCardIcon
      iconName="work"
      iconBackgroundColour={theme.palette.secondary.main}
    />
  );

  const formatter = new Intl.NumberFormat("en-AU", {
    style: "decimal",
    maximumFractionDigits: 0,
  });

  return (
    <Link
      style={{ textDecoration: "none" }}
      to={`${routes.authenticated.url}${routes.profile.url}`}
    >
      <DataCard
        title="Total Projects Running"
        dataValue={formatter.format(numProjs)}
        displayIcon={overallProjsIcon}
      />
    </Link>
  );
};

export default OverallProjects;
