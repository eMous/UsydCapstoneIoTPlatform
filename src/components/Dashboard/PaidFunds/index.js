import { useEffect } from "react";
import useTheme from "@material-ui/styles/useTheme";
import DataCard from "../components/DataCard";
import DataCardIcon from "../components/DataCardIcon";

const PaidFunds = ({ fundsAmt }) => {
  const theme = useTheme();
  const paidFundsIcon = (
    <DataCardIcon
      iconName="paid"
      iconBackgroundColour={theme.palette.secondary.main}
    />
  );

  // Formats the value into the desired readable value
  const formatter = new Intl.NumberFormat("en-AU", {
    style: "currency",
    currency: "AUD",
  });

  return (
    <DataCard
      title="Funds Paid Out"
      dataValue={formatter.format(fundsAmt)}
      displayIcon={paidFundsIcon}
    />
  );
};

export default PaidFunds;
