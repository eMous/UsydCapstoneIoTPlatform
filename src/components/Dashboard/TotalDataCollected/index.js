import useTheme from "@material-ui/styles/useTheme";
import DataCard from "../components/DataCard";
import DataCardIcon from "../components/DataCardIcon";

const TotalDataCollected = ({ numRecords }) => {
  const theme = useTheme();
  const totalDataIcon = (
    <DataCardIcon
      iconName="insights"
      iconBackgroundColour={theme.palette.secondary.main}
    />
  );

  const formatter = new Intl.NumberFormat("en-AU", {
    style: "decimal",
    maximumFractionDigits: 0,
  });

  return (
    <DataCard
      title="Data Collected (Records)"
      dataValue={`${formatter.format(numRecords)}`}
      displayIcon={totalDataIcon}
    />
  );
};

export default TotalDataCollected;
