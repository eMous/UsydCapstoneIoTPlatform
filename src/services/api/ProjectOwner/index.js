import axios from "axios";
import routes from "routes";
import { CustomException, ErrorTypes } from "utils/Error";

//Data Preview
export const DataPreview = async (projId, userIdToken) => {
  try {
    const { data, status, statusText } = await axios.post(
      routes.api.dataPreview,
      { projectId: projId },
      {
        headers: {
          Authorization: `Bearer ${userIdToken}`,
        },
      }
    );

    if (status !== 200) {
      throw new CustomException(
        " Data Preview Error",
        `Did not receive a 200 response when retrieving the API: ${statusText}`
      );
    }
    return data;
  } catch (error) {
    console.log(error);
  }
};

//Fetch the frequencies
export const FetchFrequencies = async () => {
  try {
    const { data, status, statusText } = await axios.get(
      routes.api.getSensorFrequencies
    );

    if (status !== 200) {
      throw new CustomException(
        " Frequencies Fetch Error",
        `Did not receive a 200 response when retrieving the API: ${statusText}`
      );
    }
    return data;
  } catch (error) {
    console.log(error);
  }
};

//Fetch the Api Range
export const FetchApiRange = async () => {
  try {
    const { data, status, statusText } = await axios.get(
      routes.api.getApiRange
    );

    if (status !== 200) {
      throw new CustomException(
        " API Fetch Error",
        `Did not receive a 200 response when retrieving the API: ${statusText}`
      );
    }
    return data;
  } catch (error) {
    console.log(error);
  }
};

//Fetch the sensors
export const FetchSensors = async () => {
  try {
    const { data, status, statusText } = await axios.get(routes.api.getSensors);

    if (status !== 200) {
      throw new CustomException(
        " Sensors Fetch Error",
        `Did not receive a 200 response when retrieving the sensors: ${statusText}`
      );
    }
    return data;
  } catch (error) {
    console.log(error);
  }
};

//Fetch the gender types
export const FetchGenderTypes = async () => {
  try {
    const { data, status, statusText } = await axios.get(
      routes.api.getGenderTypes
    );

    if (status !== 200) {
      throw new CustomException(
        "Gender Types Fetch Error",
        `Did not receive a 200 response when retrieving the gender types: ${statusText}`
      );
    }
    return data;
  } catch (error) {
    console.log(error);
  }
};

// Fetch the Research Field mapping
export const FetchResearchFieldMap = async () => {
  try {
    console.log("inside API call");
    const { data, status, statusText } = await axios.get(
      routes.api.getResearchFieldMap
    );

    if (status !== 200) {
      throw new CustomException(
        "Research Fields Map Error",
        `Did not receive a 200 response when retrieving the research field mapping: ${statusText}`
      );
    }

    return data;
  } catch (error) {
    console.log(error);
  }
};

// Fetch project information
export const FetchProjectInfo = async (projId, idToken) => {
  try {
    const { data, status, statusText } = await axios.post(
      routes.api.findProject,
      {
        projectId: projId,
      },
      {
        headers: {
          Authorization: "Bearer " + idToken,
        },
      }
    );

    if (status !== 200) {
      throw new CustomException(
        "Project Details Error",
        `Did not receive a 200 response when retrieving the project details: ${statusText}`,
        ErrorTypes.FETCH_PROJ_DETAILS
      );
    }

    return data;
  } catch (error) {
    return Promise.reject(error);
  }
};

//login
export const LoginApi = async (idToken) => {
  try {
    const data = await axios.post(
      routes.api.login,
      { idToken },
      {
        headers: {
          Authorization: "Bearer " + idToken,
        },
      }
    );

    if (data.status !== 200) {
      throw new CustomException(
        "login-api error",
        data.data,
        ErrorTypes.LOGIN_API
      );
    }
  } catch (error) {
    console.log(error);
  }
};

//Signout
export const SignOutApi = async (idToken) => {
  try {
    const data = await axios.post(
      routes.api.logout,
      {},
      {
        headers: {
          Authorization: "Bearer " + idToken,
        },
      }
    );

    if (data.code !== 200) {
      throw new CustomException(
        "signout-api error",
        data.errors,
        ErrorTypes.SIGNOUT_API
      );
    }
  } catch (error) {
    console.log(error);
  }
};

//get user's information for profile
export const GetUserInfoApi = async (idToken) => {
  try {
    const data = await axios.get(routes.api.profile, {
      headers: {
        Authorization: `Bearer ${idToken}`,
      },
    });
    if (data.status === 200) {
      console.log("success");
      console.log(data.data);
      return data.data;
    } else if (data.status !== 200) {
      throw new CustomException(
        "profile-api error",
        data.statusText,
        ErrorTypes.BACKEND_PROFILE
      );
    }

    return data.data || null;
  } catch (error) {
    console.log(error);
  }
};

//edit user's profile
export const editProfile = async (profileObj, userIdToken) => {
  let sanitizedProfileObj = {
    organisation: "",
    name: "",
    phoneNum: "",
    researchFields: [],
  };

  if (profileObj.organisation) {
    sanitizedProfileObj.organisation = profileObj.organisation;
  } else {
    delete sanitizedProfileObj.organisation;
  }

  if (profileObj.name) {
    sanitizedProfileObj.name = profileObj.name;
  } else {
    delete sanitizedProfileObj.name;
  }

  if (profileObj.phoneNum) {
    sanitizedProfileObj.phoneNum = profileObj.phoneNum;
  } else {
    delete sanitizedProfileObj.phoneNum;
  }

  if (profileObj.researchFields) {
    sanitizedProfileObj.researchFields = profileObj.researchFields;
  } else {
    delete sanitizedProfileObj.researchFields;
  }

  try {
    if (Object.keys(sanitizedProfileObj).length < 1) {
      throw new Error();
    }
  } catch (error) {
    const errorName = "Server-side Profile Error";
    const custErr = new CustomException(
      errorName,
      "The object provided did not meet the request body requirements.",
      ErrorTypes.BACKEND_PROFILE
    );
    console.log(custErr.toString());
  }

  try {
    await axios.post(routes.api.profile, sanitizedProfileObj, {
      headers: {
        Authorization: `Bearer ${userIdToken}`,
      },
    });
  } catch (error) {
    const errorName = "Server-side Profile Error";
    const custErr = new CustomException(
      errorName,
      "There was an issue editing the user's profile.",
      ErrorTypes.BACKEND_PROFILE
    );
    console.log(custErr.toString());
  }
};

//create a new project
export const CreateAProject = async (projectInfo, userIdToken) => {
  try {
    const data = await axios.post(routes.api.createProject, projectInfo, {
      headers: {
        Authorization: `Bearer ${userIdToken}`,
      },
    });
    if (data.status == 200) {
      return true;
    }

    if (data.status !== 200) {
      throw new CustomException(
        "project-creating-api error",
        data.statusText,
        ErrorTypes.BACKEND_PROJECT
      );
    }

    return data.data || null;
  } catch (error) {
    console.log(error);
  }
};

//update Balance
export const UpdateBalance = async (addedValue, userIdToken) => {
  try {
    console.log(addedValue);
    const data = await axios.post(routes.api.addBalance, addedValue, {
      headers: {
        Authorization: `Bearer ${userIdToken}`,
      },
    });
    if (data.status == 200) {
      return true;
    } else {
      throw new CustomException(
        "project-creating-api error",
        data.statusText,
        ErrorTypes.BACKEND_PROJECT
      );
    }
  } catch (error) {
    const errorName = "Server-side Profile Error";
    const custErr = new CustomException(
      errorName,
      "There was an issue editing the user's balance.",
      ErrorTypes.BACKEND_PROFILE
    );
    console.log(custErr.toString());
  }
};
