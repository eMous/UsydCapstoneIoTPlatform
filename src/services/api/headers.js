const protectedHeaders = {
  // 'CSRF-Token': document.querySelector('meta[property="csrf-token"]').getAttribute('content'),
  Accept: "application/json",
  "Content-Type": "application/json",
};

export { protectedHeaders as default };
