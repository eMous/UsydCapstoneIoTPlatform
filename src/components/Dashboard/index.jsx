import { useContext, useEffect, useState } from "react";
import AuthContext from "contexts/auth";
import Grid from "@material-ui/core/Grid";
import PaidFunds from "./PaidFunds";
import HoldingFunds from "./HoldingFunds";
import OverallProjects from "./OverallProjects";
import TotalDataCollected from "./TotalDataCollected";
import { GetUserInfoApi } from "services/api/ProjectOwner";

const Dashboard = () => {
  const { userToken } = useContext(AuthContext);
  const initDashboardData = {
    paidFundAmt: 0,
    holdFundAmt: 0,
    totalProjs: 0,
    totalData: 0,
  };
  const [dashboardData, setDashboardData] = useState(initDashboardData);

  /**
   * The reason for having the API calls done here instead of the individual
   * dashboard components, is because we don't have an API call specifically
   * for these statistics.
   *
   * We need to make a GET call for each of the projects the user has and
   * amalgamate the values for each statistic we want.
   *
   * To save API calls, we perform it once in this parent component, instead
   * of having each component perform their own set of API calls (repeated).
   */
  useEffect(() => {
    const fetchProjData = async () => {
      const userProfileData = await GetUserInfoApi(userToken);
      let updatedDashboardData = { ...initDashboardData, ...{} };

      if (userProfileData !== null && typeof userProfileData !== "undefined") {
        if (
          Object.hasOwnProperty.call(userProfileData, "totalHasFundedMoney")
        ) {
          updatedDashboardData.paidFundAmt =
            userProfileData.totalHasFundedMoney;
        }

        if (Object.hasOwnProperty.call(userProfileData, "balance")) {
          updatedDashboardData.holdFundAmt = userProfileData.balance;
        }

        if (Object.hasOwnProperty.call(userProfileData, "projects")) {
          updatedDashboardData.totalProjs = userProfileData.projects.length;
        }

        if (
          Object.hasOwnProperty.call(
            userProfileData,
            "totalAmountDataCollected"
          )
        ) {
          updatedDashboardData.totalData =
            userProfileData.totalAmountDataCollected;
        }

        setDashboardData(updatedDashboardData);
      }
    };

    fetchProjData();
  }, []);

  return (
    <Grid container direction="row" justify="center" spacing={3}>
      <Grid item>
        <PaidFunds fundsAmt={dashboardData.paidFundAmt} />
      </Grid>
      <Grid item>
        <HoldingFunds fundsAmt={dashboardData.holdFundAmt} />
      </Grid>
      <Grid item>
        <OverallProjects numProjs={dashboardData.totalProjs} />
      </Grid>
      <Grid item>
        <TotalDataCollected numRecords={dashboardData.totalData} />
      </Grid>
    </Grid>
  );
};

export default Dashboard;
