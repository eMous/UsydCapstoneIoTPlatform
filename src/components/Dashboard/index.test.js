import renderer from "react-test-renderer";
import Dashboard from "./index";
import { cleanup, render, screen } from "@testing-library/react";
import MockThemeProvider from "../../../__tests__/utils/mockTheme";

jest.mock("services/api/ProjectOwner");
import { GetUserInfoApi } from "services/api/ProjectOwner";

describe("Dashboard component", () => {
  test("Snapshot of Dashboard component is rendered", () => {
    const component = renderer.create(
      <MockThemeProvider>
        <Dashboard />
      </MockThemeProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("All dashboard components are rendered", () => {
    beforeEach(() => {
      // We mock the API call made by Dashboard
      GetUserInfoApi.mockImplementation(() => ({
        totalHasFundedMoney: 2000,
        holdFundAmt: 4000,
        totalAmountDataCollected: 600,
        balance: 8000,
      }));

      render(
        <MockThemeProvider>
          <Dashboard />
        </MockThemeProvider>
      );
    });

    afterEach(() => {
      cleanup();
    });

    test("Component for funds paid out is shown", () => {
      const paidFundsComponent = screen.getByText(/funds paid out/i);
      expect(paidFundsComponent).toBeInTheDocument();
    });

    test("Component for overall projects is shown", () => {
      const overallProjsComponent = screen.getByText(/total projects running/i);
      expect(overallProjsComponent).toBeInTheDocument();
    });

    test("Component for holding funds is shown", () => {
      const holdingFundsComponent = screen.getByText(/funds in holding/i);
      expect(holdingFundsComponent).toBeInTheDocument();
    });

    test("Component for total data collected is shown", () => {
      const dataCollectedComponent = screen.getByText(/data collected/i);
      expect(dataCollectedComponent).toBeInTheDocument();
    });
  });
});
