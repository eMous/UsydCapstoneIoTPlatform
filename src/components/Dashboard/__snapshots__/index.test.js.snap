// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component Snapshot of Dashboard component is rendered 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-justify-xs-center"
>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root"
      >
        <div
          className="MuiCardHeader-avatar"
        >
          <div
            className="MuiAvatar-root MuiAvatar-rounded makeStyles-rounded-1 MuiAvatar-colorDefault"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              paid
            </span>
          </div>
        </div>
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            Funds Paid Out
          </span>
          <span
            className="MuiTypography-root MuiCardHeader-subheader MuiTypography-h4 MuiTypography-colorPrimary MuiTypography-displayBlock"
          >
            A$0.00
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root"
      >
        <div
          className="MuiCardHeader-avatar"
        >
          <div
            className="MuiAvatar-root MuiAvatar-rounded makeStyles-rounded-2 MuiAvatar-colorDefault"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              account_balance
            </span>
          </div>
        </div>
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            Funds in Holding (To be paid)
          </span>
          <span
            className="MuiTypography-root MuiCardHeader-subheader MuiTypography-h4 MuiTypography-colorPrimary MuiTypography-displayBlock"
          >
            A$0.00
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root"
      >
        <div
          className="MuiCardHeader-avatar"
        >
          <div
            className="MuiAvatar-root MuiAvatar-rounded makeStyles-rounded-3 MuiAvatar-colorDefault"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              work
            </span>
          </div>
        </div>
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            Total Projects Running
          </span>
          <span
            className="MuiTypography-root MuiCardHeader-subheader MuiTypography-h4 MuiTypography-colorPrimary MuiTypography-displayBlock"
          >
            0
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root"
      >
        <div
          className="MuiCardHeader-avatar"
        >
          <div
            className="MuiAvatar-root MuiAvatar-rounded makeStyles-rounded-4 MuiAvatar-colorDefault"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              insights
            </span>
          </div>
        </div>
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            Data Collected (Records)
          </span>
          <span
            className="MuiTypography-root MuiCardHeader-subheader MuiTypography-h4 MuiTypography-colorPrimary MuiTypography-displayBlock"
          >
            0
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
